src/components/ProductFilter.js

```diff
@@ -0,0 +1,278 @@
+import { useEffect, useMemo, useReducer } from "react";
+import { useState } from "react";
+import "./css/ProductFilter.css";
+
+const ProductFilter = ({ setFilter, className, buttonFun, data }) => {
+  //ready to get irritated as i got seeing my own code why wont you
+
+  const [state, dispatch] = useReducer(FilterReducer, {
+    sort: null,
+    onlyInStock: false,
+    onlyFastDelivery: false,
+    priceRange: 0,
+    brand: [],
+    rating: 0,
+  });
+  const brands = data.brands;
+
+  useEffect(() => {
+    //pass up the latest filter configuration to its parent component
+    setFilter(state);
+  }, [state, setFilter]);
+
+  return (
+    <div className={`productFilter ${className ?? ""}`}>
+      <div className="productFilter_heading">
+        <h4>Filter</h4>
+        <button
+          className="productFilter_clear"
+          onClick={() => {
+            dispatch({ type: "reset" });
+            if (buttonFun) buttonFun();
```

Comment:
Can improve the naming ... buttonFun does not indicates what the function will do